lib/block_view_atomic_txns.go

```diff
@@ -0,0 +1,476 @@
+	for _, txn := range txnData.Txns {
+		txnBytes, err := txn.ToBytes(preSignature)
+		if err != nil {
+			return nil, errors.Wrap(err,
+				"AtomicTxnsWrapperMetadata.ToBytes: Problem serializing txn")
+		}
+		data = append(data, UintToBuf(uint64(len(txnBytes)))...)
+		data = append(data, txnBytes...)
+	}
+	return data, nil
+}
+
+func (txnData *AtomicTxnsWrapperMetadata) FromBytes(data []byte) error {
+	rr := bytes.NewReader(data)
+
+	// Read the number of transactions within the atomic transaction.
+	numTxns, err := ReadUvarint(rr)
+	if err != nil {
+		return errors.Wrap(err,
+			"AtomicTxnsWrapperMetadata.FromBytes: Problem reading numTxns")
+	}
+	txnData.Txns, err = SafeMakeSliceWithLength[*MsgDeSoTxn](numTxns)
+	if err != nil {
+		return errors.Wrap(err, "AtomicTxnsWrapperMetadata.FromBytes: Problem allocating txnData.Txns")
+	}
+
+	// Read the transactions.
+	for ii := uint64(0); ii < numTxns; ii++ {
+		txnData.Txns[ii] = &MsgDeSoTxn{}
+
+		// Figure out how many bytes are associated with the ith transaction.
+		numTxnBytes, err := ReadUvarint(rr)
+		if err != nil {
+			return errors.Wrap(err,
+				"AtomicTxnsWrapperMetadata.FromBytes: Problem reading number of bytes in transaction")
+		}
+
+		// Allocate memory for the transaction bytes to be read into.
+		txnBytes, err := SafeMakeSliceWithLength[byte](numTxnBytes)
+		if err != nil {
+			return errors.Wrap(err,
+				"AtomicTxnsWrapperMetadata.FromBytes: Problem allocating bytes for transaction")
+		}
+
+		// Read the transaction into the txnBytes memory buffer.
+		if _, err = io.ReadFull(rr, txnBytes); err != nil {
+			return errors.Wrap(err,
+				"AtomicTxnsWrapperMetadata.FromBytes: Problem reading bytes for transaction")
+		}
+
+		// Convert the txnBytes buffer to a MsgDeSoTxn struct.
+		if err = txnData.Txns[ii].FromBytes(txnBytes); err != nil {
+			return errors.Wrap(err,
+				"AtomicTxnsWrapperMetadata.FromBytes: Problem parsing transaction bytes")
+		}
+	}
+	return nil
+}
+
+func (txnData *AtomicTxnsWrapperMetadata) New() DeSoTxnMetadata {
+	return &AtomicTxnsWrapperMetadata{}
+}
+
+//
+// HELPER FUNCTIONS: MsgDeSoTxn
+//
+
+func (msg *MsgDeSoTxn) IsAtomicTxnsInnerTxn() bool {
+	// An atomic transaction is qualified by the existence of the NextAtomicTxnPreHash
+	// and PreviousAtomicTxnPreHash keys in the ExtraData map.
+	if _, keyExists := msg.ExtraData[NextAtomicTxnPreHash]; !keyExists {
+		return false
+	}
+	if _, keyExists := msg.ExtraData[PreviousAtomicTxnPreHash]; !keyExists {
+		return false
+	}
+	return true
+}
+
+func (msg *MsgDeSoTxn) AtomicHash() (*BlockHash, error) {
+	// Create a duplicate of the transaction to ensure we don't edit the existing transaction.
+	msgDuplicate, err := msg.Copy()
+	if err != nil {
+		return nil, errors.Wrap(err, "MsgDeSoTxn.AtomicHash: Cannot create duplicate transaction")
+	}
+
+	// Sanity check that the transaction includes the necessary extra data to be included in an atomic transaction.
+	if !msgDuplicate.IsAtomicTxnsInnerTxn() {
```

Comment:
@poolcoke do we want to do this check in AtomicHash()? Feels like we should probably remove it. I could see us wanting to compute the atomic hash for a txn BEFORE we've set its previous/next hashes in transaction construction. Wdyt?